frame/contracts/src/rent.rs

```diff
@@ -127,75 +177,200 @@ fn try_evict_or_and_pay_rent<T: Trait>(
+	let alive_contract_info = match contract_info {
+		None | Some(ContractInfo::Tombstone(_)) => return contract_info,
+		Some(ContractInfo::Alive(contract)) => contract,
+	};
+
+	let current_block_number = <frame_system::Module<T>>::block_number();
+	let verdict = consider_case::<T>(
+		account,
+		current_block_number,
+		Zero::zero(),
+		&alive_contract_info,
+	);
+	enact_verdict(account, alive_contract_info, current_block_number, verdict)
 }
 
-/// Evict the account if it should be evicted at the given block number.
+/// Process a snitch that a contract under the given address should be evicted.
+///
+/// Enact the eviction right away if the contract should be evicted and return true.
+/// Otherwise, **do nothing** and return false.
 ///
-/// `handicap` gives a way to check or pay the rent up to a moment in the past instead
+/// The `handicap` parameter gives a way to check the rent to a moment in the past instead
 /// of current block. E.g. if the contract is going to be evicted at the current block,
-/// `handicap=1` can defer the eviction for 1 block.
+/// `handicap = 1` can defer the eviction for 1 block. This is useful to handicap certain snitchers
+/// relative to others.
 ///
-/// NOTE: This function acts eagerly.
-pub fn try_evict<T: Trait>(account: &T::AccountId, handicap: T::BlockNumber) -> RentOutcome {
-	try_evict_or_and_pay_rent::<T>(account, handicap, false).0
+/// NOTE this function performs eviction eagerly. All changes are read and written directly to
+/// storage.
+pub fn snitch_contract_should_be_evicted<T: Trait>(
+	account: &T::AccountId,
+	handicap: T::BlockNumber,
+) -> bool {
+	let contract_info = <ContractInfoOf<T>>::get(account);
+	let alive_contract_info = match contract_info {
+		None | Some(ContractInfo::Tombstone(_)) => return false,
+		Some(ContractInfo::Alive(contract)) => contract,
+	};
+	let current_block_number = <frame_system::Module<T>>::block_number();
+	let verdict = consider_case::<T>(
+		account,
+		current_block_number,
+		handicap,
+		&alive_contract_info,
+	);
+
+	// Enact the verdict only if the contract gets removed.
+	match verdict {
+		Verdict::Kill | Verdict::Evict { .. } => {
+			enact_verdict(account, alive_contract_info, current_block_number, verdict);
+			true
+		}
+		_ => false,
+	}
+}
+
+/// Returns the projected time a given contract will be able to sustain paying its rent. The
+/// returned projection is relevent for the current block, i.e. it is as if the contract was
+/// accessed at the beginning of the current block. Returns `None` in case if the contract was
+/// evicted before or as a result of the rent collection.
+///
+/// The returned value is only an estimation. It doesn't take into account any top ups, changing the
+/// rent allowance, or any problems coming from withdrawing the dues.
+///
+/// NOTE that this is not a side-effect free function! It will actually collect rent and then
+/// compute the projection.
+pub fn compute_rent_projection<T: Trait>(
```

Comment:
Maybe call it `compute_and_collect_rent`?

Reply:
`compute_and_collect_rent_projection` you mean? Yeah, that sounds good.

Reply:
ah, no, we compute rent projection but collect rent. That the reason why I didn't like `compute_and_collect_rent`, but `compute_and_collect_rent_projection` is no better.

Reply:
Right, I see what you mean. I guess `collect_rent_and_get_projection` would be the right name, but it becomes quite longish.

Now it seems we have a inconspicuous name for a function with quite important side effects. I think we could flip it and have a more side-effecty name that happens to return some additional value. What I have in mind is `fn collect_rent() -> FutureRentProjection` function to emphasize on the side effect part, not the return value part.

Reply:
Yeah, but the problem that we already have `collect_rent` which is actually used in on-chain logic...

This one is used only for the RPC, the difference from `collect_rent` is that this one returns additional information which is useless in the on-chain version.

Reply:
I see, thanks for explanations, sounds good to leave it as is then. Perhaps a comment indicating that the function is used only for `RuntimeApi` calls, which would not persist the changes anyway would be good.